app/models/tag_implication.rb

```diff
@@ -1,24 +1,39 @@
+# frozen_string_literal: true
+
 class TagImplication < ApplicationRecord
   before_create :validate_uniqueness
 
   def validate_uniqueness
-    if self.class.exists? ["(predicate_id = ? AND consequent_id = ?) OR (predicate_id = ? AND consequent_id = ?)", predicate_id, consequent_id, consequent_id, predicate_id]
-      errors.add(:base, "Tag implication already exists")
-      throw :abort
-    end
+    return unless self.class.exists? ['(predicate_id = ? AND consequent_id = ?) OR (predicate_id = ? AND consequent_id = ?)', predicate_id, consequent_id, consequent_id, predicate_id]
+
+    errors.add(:base, 'Tag implication already exists')
+    throw :abort
   end
 
   # Destroys the alias and sends a message to the alias's creator.
   def destroy_and_notify(current_user, reason)
     if creator_id && creator_id != current_user.id
       msg = "A tag implication you submitted (#{predicate.name} → #{consequent.name}) was deleted for the following reason: #{reason}."
 
-      Dmail.create(:from_id => current_user.id, :to_id => creator_id, :title => "One of your tag implications was deleted", :body => msg)
+      Dmail.create(from_id: current_user.id, to_id: creator_id, title: 'One of your tag implications was deleted', body: msg)
     end
 
     destroy
   end
 
+  def self.retrieve_matching_tag_id
```

Comment:
it's already mixed up but don't further randomly mix up class and instance method. Class methods should be defined before instance methods

Reply:
Class methods have been defined first